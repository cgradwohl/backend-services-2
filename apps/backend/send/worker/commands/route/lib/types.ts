import { IProviderConfiguration } from "~/send/types";

export type MatchingProvidersByChannel = {
  [channel: string]: IProviderConfiguration[];
};
