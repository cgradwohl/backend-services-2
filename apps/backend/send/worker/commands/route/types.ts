import { MessageStatusReason } from "~/lib/message-service/types";
import {
  IChannelConfiguration,
  IChannelProvider,
  IChannelProviderConfiguration,
  IConditionalConfig,
} from "~/types.api";

export interface IChannelSummary {
  channel: string;
  reason: string;
  selected: boolean;
}

export interface IUndeliverableEventLog {
  type: MessageStatusReason;
  reason: string;
  data: Record<string, unknown>;
}

export interface IRoutingSummary {
  channel: string;
  provider?: string;
  selected: boolean;
  timedout?: boolean;
  blockIds?: string[] | null;
  config?: IChannelConfiguration | IChannelProviderConfiguration | undefined;
  id?: string;
  reason?: string;
  providers?: IChannelProvider[] | null;
  taxonomy?: string;
  disabled?: boolean;
  label?: string;
  conditional?: IConditionalConfig;
  key?: string;
  // Provider Configuration ID
  configurationId?: string;
}
