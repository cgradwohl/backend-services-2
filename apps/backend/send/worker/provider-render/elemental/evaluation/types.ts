import { ElementalNodeIR } from "~/api/send/types";

export type ElementalRefs = { [refId: string]: ElementalNodeIR };
