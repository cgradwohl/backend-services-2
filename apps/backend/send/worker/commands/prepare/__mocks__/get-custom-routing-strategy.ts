export const getUserRouting = jest.fn();
