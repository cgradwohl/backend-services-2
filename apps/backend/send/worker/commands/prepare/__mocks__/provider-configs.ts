export const mockProviderConfigs: any = [
  {
    json: {
      domain: "sandboxb1d4dd0eb044497987e3c14f3c7c4db6.mailgun.org",
      apiKey: "4cea38609bdedcd992c18fd595a55a6d-7b8c9ba8-91e9abcd",
      fromAddress: "cmoney@example.com",
      provider: "apn",
    },
    id: "f2661ebd-dd91-417c-8b37-a9870fd0cbf4",
  },
  {
    json: {
      domain: "sandboxb1d4dd0eb044497987e3c14f3c7c4db6.mailgun.org",
      apiKey: "4cea38609bdedcd992c18fd595a55a6d-7b8c9ba8-91e9abcd",
      fromAddress: "cmoney@example.com",
      provider: "firebase-fcm",
    },
    id: "fsdkjhf-fdsf-sdf-sdfs-sdffsdf",
  },
  {
    json: {
      apiKey: "SG.gf.df",
      fromAddress: "chris@courier.com",
      checkDeliveryStatus: false,
      provider: "airship",
    },
    id: "a0e317ae-b30b-482f-a503-5e772e234f65",
  },
  {
    json: {
      accountSid: "wetwetgs",
      authToken: "tewthfdadfdf",
      messagingServiceSid: "greswetwtre",
      provider: "twilio",
    },
    id: "7593c6b6-f241-47f0-85b4-b5386cd60086",
  },
  {
    json: {
      accountSid: "wetwetgs",
      authToken: "tewthfdadfdf",
      messagingServiceSid: "greswetwtre",
      provider: "vonage",
    },
    id: "7593c6b6-f241-47f0-85b4-vonage",
  },
  {
    json: {
      accountSid: "wretewrtwq",
      authToken: "wtewetsgfdgssdgf",
      messagingServiceSid: "sdfgsdfgfdsdssdfg",
      provider: "mailjet",
    },
    id: "7593c6b6-f241-47f0-85b4-b5386cd600er6",
  },
  {
    json: {
      accountSid: "GSGDSewrewrewr",
      authToken: "gdfsgferger345sfa",
      messagingServiceSid: "sdfsggfdsdfsgfsdfger",
      provider: "sendgrid",
    },
    id: "2321",
  },
  {
    json: {
      accountSid: "gsdfds4353454",
      authToken: "324323423423",
      messagingServiceSid: "adsfasadfaswewr",
      provider: "slack",
    },
    id: "slack2e1231",
  },
];
