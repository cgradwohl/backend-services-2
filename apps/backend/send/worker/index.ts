import { createEventHandlerWithFailures } from "~/lib/kinesis/create-event-handler";
import { actionCommands } from "./commands";
import { IAction } from "../types";
import { CourierLogger } from "~/lib/logger";
import { NonRetryableSendError } from "../errors/types";
import { retryMessage } from "../utils/retry-message";
import { sendErrorMetric } from "~/lib/courier-emf/logger-metrics-utils";
const { logger } = new CourierLogger("ActionWorkerLogger");

async function worker(action: IAction) {
  try {
    const command = actionCommands[action.command];

    await command(action);
  } catch (error: unknown) {
    logger.error(`:: ${action?.command} COMMAND ERROR ::`, {
      action,
      error,
      tenantId: action?.tenantId,
      traceId: action?.requestId,
    });

    await sendErrorMetric({
      action,
      error,
      tenantId: action?.tenantId,
      traceId: action?.requestId,
    });

    // NOTE: This limits the scope of the new Error Retry Policy.
    // TODO: Remove after all command errors have been classified
    if (
      [
        "accept",
        "ad-hoc-list",
        "list",
        "list-pattern",
        "request",
        "route",
        "send-audiences",
        "send-audiences-member",
      ].includes(action?.command)
    ) {
      // an explicit non retryable error was thrown, so don't retry
      if (error instanceof NonRetryableSendError) {
        return;
      }

      await retryMessage({
        ...action,
        streamName: process.env.ACTION_STREAM!,
      });
    }
  }
}

export default createEventHandlerWithFailures<IAction>(
  worker,
  process.env.ACTION_STREAM_SEQUENCE_TABLE
);
