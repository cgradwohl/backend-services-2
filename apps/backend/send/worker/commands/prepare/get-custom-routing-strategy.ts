import { MessageRouting } from "~/api/send/types";
import { toUuid } from "~/lib/api-key-uuid";
import { preferenceTemplateService } from "~/preferences/services/dynamo-service";
import { IPreference } from "~/types.public";

/**
 * Get's a user provided routing strategy. Only used by get-routing, which is the function
 * you are probably looking for. Not this one.
 */
export async function getUserRouting({
  userId,
  tenantId,
  preferenceTemplateId,
}: {
  userId?: string;
  tenantId: string;
  preferenceTemplateId?: string;
}): Promise<MessageRouting | undefined> {
  if (!userId || !preferenceTemplateId) {
    return undefined;
  }

  const preferenceTemplateUuId = toUuid(preferenceTemplateId);

  const resourceId = `${userId}#${preferenceTemplateUuId}`;

  const recipientPreferencesIfExists = (
    await preferenceTemplateService(tenantId, "").get<{
      value: IPreference;
    }>("recipients", resourceId)
  )?.value;

  const shouldUseExistingStrategy =
    !recipientPreferencesIfExists?.hasCustomRouting;

  if (shouldUseExistingStrategy) return undefined;

  const routingPreferencesForCurrentUser =
    recipientPreferencesIfExists?.routingPreferences;

  return {
    method: "all",
    channels: [...routingPreferencesForCurrentUser],
  };
}
