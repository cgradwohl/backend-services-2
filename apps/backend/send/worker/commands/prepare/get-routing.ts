import { Message, UserRecipient } from "~/api/send/types";
import {
  generateRouting,
  generateTemplateV1Routing,
  RouteNode,
  RoutingStrategy,
} from "~/lib/send-routing";
import { TokensByProvider } from "~/lib/token-storage";
import { IProviderConfiguration, ISendMessageContext } from "~/send/types";
import { INotificationWire, IProfile, NotificationCategory } from "~/types.api";
import { IProfilePreferences } from "~/types.public";
import { getUserRouting } from "./get-custom-routing-strategy";

export interface GetRoutingTreeOpts {
  message: Message;
  tenantId: string;
  templateV1?: INotificationWire;
  profile: IProfile;
  tokens?: TokensByProvider;
  providerConfigs: IProviderConfiguration[];
  strategy: RoutingStrategy;

  // Template V1 stuff
  category?: NotificationCategory;
  preferences?: Partial<IProfilePreferences>;
  variableData: ISendMessageContext["variableData"];
}

/** To maintain backwards compatibility with pending calls the result of this function is intended to be spread onto ISendMessageContext */
export const getRoutingTree = async (
  opts: GetRoutingTreeOpts
): Promise<RoutingContext> => {
  const { tenantId, templateV1 } = opts;

  const userRouting = await getUserRouting({
    userId: (opts.message.to as UserRecipient)?.user_id,
    tenantId,
    preferenceTemplateId: opts.templateV1?.json?.preferenceTemplateId,
  });

  const routingOverride = userRouting ?? opts.message.routing;

  if (templateV1 && !routingOverride) {
    return {
      routingTree: await generateTemplateV1Routing({
        category: opts.category,
        templateV1,
        preferences: opts.preferences,
        variableData: opts.variableData,
        tokens: opts.tokens,
        providers: opts.providerConfigs,
      }),
    };
  }

  const strategy = {
    ...opts.strategy,
    routing: routingOverride ?? opts.strategy.routing,
  };

  return {
    routingTree: await generateRouting({
      providerConfigs: opts.providerConfigs,
      templateV1: opts.templateV1,
      tokens: opts.tokens,
      params: { profile: opts.profile, data: opts.message.data },
      strategy,
    }),
  };
};

export interface RoutingContext {
  routingTree: RouteNode;
}
