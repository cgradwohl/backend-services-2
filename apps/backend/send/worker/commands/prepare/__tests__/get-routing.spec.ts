import { Message } from "~/api/send/types";
import { getUserRouting } from "../get-custom-routing-strategy";
import { getRoutingTree, GetRoutingTreeOpts } from "../get-routing";
import { mockProviderConfigs } from "../__mocks__/provider-configs";

jest.mock("../get-custom-routing-strategy");
jest.mock("~/lib/get-environment-variable");
jest.mock("~/lib/sentry");

describe("get routing tree", () => {
  beforeEach(jest.clearAllMocks);

  const baseMessage: Message = {
    content: { body: "hello" },
    to: {},
  };

  const baseOpts: GetRoutingTreeOpts = {
    message: baseMessage,
    tenantId: "tenant-id",
    profile: {},
    providerConfigs: mockProviderConfigs,
    strategy: {
      routing: { method: "all", channels: ["sms"] },
      channels: {},
      providers: {},
    },
    variableData: {
      profile: {},
      data: {},
    } as any,
  };

  it("calls getUserRouting", async () => {
    expect.assertions(1);
    await getRoutingTree(baseOpts);
    expect(getUserRouting).toHaveBeenCalled();
  });

  it("returns the standard supplied strategy when routing tree is not enabled", async () => {
    expect.assertions(1);
    const routing = await getRoutingTree(baseOpts);
    expect((routing as any).strategy).toMatchObject(baseOpts.strategy);
  });

  it("returns empty object when routing tree is not enabled, template is passed, and no message.routing is provided", async () => {
    expect.assertions(1);
    const routing = await getRoutingTree({
      ...baseOpts,
      message: { template: "", to: {} },
      templateV1: {} as any,
    });
    expect(routing).toMatchObject({});
  });

  it("returns a routing tree when enabled for tenant", async () => {
    expect.assertions(1);
    const routing = await getRoutingTree({
      ...baseOpts,
      tenantId: "038ba22e-e641-4a56-a1bf-463c81c65ef2",
    });
    expect(routing).toMatchSnapshot();
  });
});
