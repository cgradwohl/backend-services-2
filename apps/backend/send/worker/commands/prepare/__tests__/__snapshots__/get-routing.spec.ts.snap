// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`get routing tree returns a routing tree when enabled for tenant 1`] = `
Object {
  "routingTree": Object {
    "address": Array [],
    "nodes": Array [
      Object {
        "address": Array [
          0,
        ],
        "nodes": Array [
          Object {
            "address": Array [
              0,
              0,
            ],
            "channel": "sms",
            "failureReason": "Information required by the provider was not included.",
            "failureType": "MISSING_PROVIDER_SUPPORT",
            "provider": "twilio",
            "type": "dead-leaf",
          },
          Object {
            "address": Array [
              0,
              1,
            ],
            "channel": "sms",
            "failureReason": "Information required by the provider was not included.",
            "failureType": "MISSING_PROVIDER_SUPPORT",
            "provider": "vonage",
            "type": "dead-leaf",
          },
        ],
        "type": "branch",
      },
    ],
    "type": "branch",
  },
  "strategy": Object {
    "channels": Object {},
    "providers": Object {},
    "routing": Object {
      "channels": Array [
        "sms",
      ],
      "method": "all",
    },
  },
}
`;
